crates/burn-autodiff/src/runtime/graph.rs

```diff
@@ -0,0 +1,285 @@
+                    self.register_key(graph.origin, node);
+                }
+
+                graph
+            }
+            GraphAnalysis::Collisions(graphs) => self.merge(node, graphs),
+        }
+    }
+
+    /// Analyses the graph for a given node and its parents, returning the associated `GraphAnalysis`.
+    fn analyse(&mut self, node: NodeId, parents: &[Parent]) -> GraphAnalysis {
+        // If no parents, there is no collision, therefore a single graph is ok.
+        if parents.is_empty() {
+            let graph = match self.graphs.get(&node) {
+                Some(val) => val.clone(),
+                None => self.new_graph(node),
+            };
+            return GraphAnalysis::NoCollision(graph);
+        };
+
+        // We collect all graphs of parents and of the current node based on their origin node id.
+        let mut graphs = HashMap::<NodeId, Arc<Graph>>::new();
+
+        if let Some(val) = self.graphs.get(&node) {
+            graphs.insert(val.origin, val.clone());
+        }
+
+        for parent in parents {
+            match self.graphs.get(&parent.id) {
+                Some(graph) => graphs.insert(graph.origin, graph.clone()),
+                None => continue,
+            };
+        }
+
+        if graphs.is_empty() {
+            return match self.graphs.get(&node) {
+                Some(old) => GraphAnalysis::NoCollision(old.clone()),
+                None => GraphAnalysis::NoCollision(self.new_graph(node)),
+            };
+        }
+
+        if graphs.len() == 1 {
+            return GraphAnalysis::NoCollision(graphs.drain().next().unwrap().1);
+        }
+
+        GraphAnalysis::Collisions(graphs)
+    }
+
+    /// Merges multiple graphs associated with a node into a single graph.
+    fn merge(&mut self, node: NodeId, mut graphs: HashMap<NodeId, Arc<Graph>>) -> Arc<Graph> {
+        let mut graphs = graphs.drain().map(|g| g.1);
+
+        let main = graphs.next().expect("At least one graph");
+        self.register_key(main.origin, node);
+
+        let mut state = main.state.lock().unwrap();
+
+        for graph in graphs {
+            self.merge_two(&mut state, &main, graph);
+        }
+
+        self.graphs.insert(main.origin, main.clone());
+        self.graphs.insert(node, main.clone());
+
+        core::mem::drop(state);
+
+        main
+    }
+
+    /// Registers a key for a given origin node.
+    fn register_key(&mut self, origin: NodeId, key: NodeId) {
+        if !self.keys.contains_key(&origin) {
+            // Ensure an entry exists for this origin
+            self.keys.insert(origin, HashSet::new());
+        }
+
+        if origin != key {
+            // Register this node to point to the origin graph
+            self.keys.get_mut(&origin).unwrap().insert(key);
+        }
+    }
+
+    /// Merges two graphs by combining their states and updating graph mappings.
+    fn merge_two(&mut self, main_state: &mut GraphState, main: &Arc<Graph>, merged: Arc<Graph>) {
```

Comment:
In this function, should the `merged`'s `NodeId` be inserted into the `main`'s key set? (assuming this would be necessary, and `merged` wasn't inserted as a key of itself)